Allow Devise 4.x

connorshea opened the pull request:
This PR includes the following changes:
- Allow Devise `>= 3.5.0 `.
- Update the README to include `devise_parameter_sanitizer` syntax for Devise 4.0.0, see the [4.0.0 Changelog](https://github.com/plataformatec/devise/blob/master/CHANGELOG.md#400rc1---2016-01-02) for the specific changes.

This resolves #58.

I've tested this with Devise 4.0.0.rc2 on our Rails 5 branch for ImageHex, and there are some deprecation warnings, but everything works just fine. :+1:

connorshea commented:
Hm, it looks like we may need to use appraisals for the test suite. Devise 4.0.0 doesn't support Ruby 1.9 or 2.0, so we'd need to use 3.x for Ruby 1.9/2.0.

Alternatively, we can bump to `devise-two-factor` 2.0 and drop support for Ruby 1.9/2.0 since support for both has been [dropped](https://www.ruby-lang.org/en/news/2016/02/24/support-plan-of-ruby-2-0-0-and-2-1/). This may be the preferable choice.

connorshea commented:
Updating to attr_encrypted 2.x, as requested in #69, would also require dropping 1.9 as they require `>= 2.0.0`. I'll update this PR to include a requirement for `>= 2.1.0` and drop 1.9/2.0 from the Travis builds. Otherwise these changes probably won't work.

connorshea commented:
This is the problem I'm seeing with attr_encrypted 2.0.0 (for whatever reason on prior versions I think it was intercepting method_missing errors and claiming them as its own, which is why Travis claims it's attr_encrypted's fault.



```
Connors-MacBook-Pro:devise-two-factor connor$ bundle exec rake
/Users/connor/.rvm/rubies/ruby-2.1.8/bin/ruby -I/Users/connor/.rvm/gems/ruby-2.1.8/gems/rspec-core-3.4.4/lib:/Users/connor/.rvm/gems/ruby-2.1.8/gems/rspec-support-3.4.1/lib /Users/connor/.rvm/gems/ruby-2.1.8/gems/rspec-core-3.4.4/exe/rspec spec/devise/models/two_factor_authenticatable_spec.rb spec/devise/models/two_factor_backupable_spec.rb
/Users/connor/.rvm/gems/ruby-2.1.8/gems/devise-3.5.6/lib/devise/models/database_authenticatable.rb:30:in `block in <module:DatabaseAuthenticatable>': undefined method `after_update' for TwoFactorAuthenticatableDouble:Class (NoMethodError)
```

Also it doesn't help that rake updated to v11 today, breaking rspec without a deprecation warning :)

Here's the APIDock page for `after_update`: http://apidock.com/rails/ActiveRecord/Callbacks/after_update

It was deprecated after 2.3.8, so I have no idea why we'd be using it.

connorshea commented:
Devise 3.5.6 is the latest version of devise, but for whatever reason it's the one using `after_update`. I have no idea why. So, that's interesting... Digging more into it now.

connorshea commented:
Okay. Well. Apparently it works on Devise 4.5.2 but not 4.5.3 or above. Also attr_encrypted 2.0 must have removed `attr_encrypted?` at some point :(

connorshea commented:
I [opened an issue](https://github.com/attr-encrypted/attr_encrypted/issues/198) for `attr_encrypted?` not working in attr_encrypted 2.0.0. Hopefully we can get that resolved/replaced.

For now I've made it use 1.4.0 and Devise 4.5.2. I'll open an issue for Devise as well now. All the tests pass locally, so Travis *should* pass.

connorshea commented:
@bsedat after Devise (hopefully) fixes the issue, or we figure out how to work around it somehow, I'll loosen the Devise requirements and then squash my commits so there won't be a mess in the commit log :)

bsedat commented:
Thank you so much again for getting the CI fixed and dependencies updated!

connorshea commented:
@bsedat you're welcome :) Any tips on getting [this](https://github.com/plataformatec/devise/issues/3992#issuecomment-194597250) working? I can't seem to figure out where it should go. :sweat_smile:

bsedat commented:
Did Lucas Mazza's suggestion to add `define_model_callbacks :update` to [`TwoFactorAuthenticatableDouble`](https://github.com/tinfoil/devise-two-factor/blob/e59657522142bb89df95b9f4862122876a566d04/spec/devise/models/two_factor_authenticatable_spec.rb#L4-L16) fix it or change the error?

connorshea commented:
@bsedat it changed the error pretty much no matter where I put it, or just gave me the same error when I run RSpec.

```ruby
  include ::ActiveModel::Validations::Callbacks
  extend  ::Devise::Models
  define_method_callbacks :update
```

This gets me:
```
/Users/connor/Programming/devise-two-factor/spec/devise/models/two_factor_authenticatable_spec.rb:7:in '<class:TwoFactorAuthenticatableDouble>': undefined method 'define_method_callbacks' for TwoFactorAuthenticatableDouble:Class (NoMethodError)
```
and suggests that I may have meant `define_callbacks` instead.

```ruby
  include ::ActiveModel::Validations::Callbacks
  extend  ::Devise::Models
  define_callbacks :update
```

The above gets me 

```
/Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models/database_authenticatable.rb:30:in `block in <module:DatabaseAuthenticatable>': undefined method `after_update' for TwoFactorAuthenticatableDouble:Class (NoMethodError)
```

This is with my latest commit to my fork, the aforementioned changes in this comment, and also changing the runtime dependencies for attr_encrypted and devise to the following:

```ruby
s.add_runtime_dependency 'attr_encrypted'
s.add_runtime_dependency 'devise',         '4.0.0.rc2'
```

And then running bundle install.

bsedat commented:
I think you also need to `extend ActiveModel::Callbacks` in the `TwoFactorAuthenticatableDouble` to make the method available. We use a class that isn't persisted via ActiveRecord so we need to opt-into the pieces that normally exist by subclassing `ActiveRecord::Base`. Some of the other functionality is listed [here](http://guides.rubyonrails.org/active_model_basics.html#callbacks).

So it would be something like:
```ruby
extend ::ActiveModel::Callbacks
include ::ActiveModel::Validations::Callbacks
extend  ::Devise::Models

define_model_callbacks :update
```

connorshea commented:
@bsedat for whatever reason that causes the `attr_encrypted?` issue to re-appear.

```
/Users/connor/Programming/devise-two-factor/lib/devise_two_factor/models/two_factor_authenticatable.rb:12:in `block in <module:TwoFactorAuthenticatable>': undefined method `attr_encrypted?' for TwoFactorAuthenticatableDouble:Class (NoMethodError)
```

Downgrading to attr_encrypted 1.4.0 causes the after_update issue again.

```
/Users/connor/.rvm/gems/ruby-2.3.0/gems/attr_encrypted-1.4.0/lib/attr_encrypted.rb:259:in `method_missing': undefined method `after_update' for TwoFactorBackupableDouble:Class (NoMethodError)
```

Thank you for the help, though. Would you like me to push up my changes so you can play around with them?

bsedat commented:
Looks like newer versions of attr_encrypted don't monkey-patch `Object` anymore (probably a good thing!). We'll need to `extend AttrEncrypted` in the Double class as well.

connorshea commented:
@bsedat I added the extend as you suggested and updated attr_encrypted to 2.0.0.

```ruby
class TwoFactorAuthenticatableDouble
  extend ::ActiveModel::Callbacks
  extend AttrEncrypted
  include ::ActiveModel::Validations::Callbacks
  extend  ::Devise::Models

  define_model_callbacks :update
```

Here's the error when running rspec (with stack trace this time):

```
/Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models/database_authenticatable.rb:30:in `block in <module:DatabaseAuthenticatable>': undefined method `after_update' for TwoFactorBackupableDouble:Class (NoMethodError)
Did you mean?  after_validation
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/concern.rb:120:in `class_eval'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/concern.rb:120:in `append_features'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/concern.rb:117:in `include'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/concern.rb:117:in `block in append_features'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/concern.rb:117:in `each'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/concern.rb:117:in `append_features'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models.rb:103:in `include'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models.rb:103:in `block (2 levels) in devise'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models.rb:87:in `each'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models.rb:87:in `block in devise'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models.rb:114:in `devise_modules_hook!'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/devise-4.0.0.rc2/lib/devise/models.rb:84:in `devise'
	from /Users/connor/Programming/devise-two-factor/spec/devise/models/two_factor_backupable_spec.rb:7:in `<class:TwoFactorBackupableDouble>'
	from /Users/connor/Programming/devise-two-factor/spec/devise/models/two_factor_backupable_spec.rb:3:in `<top (required)>'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:268:in `load'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:268:in `block in load'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:268:in `load'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/configuration.rb:1361:in `block in load_spec_files'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/configuration.rb:1359:in `each'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/configuration.rb:1359:in `load_spec_files'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:106:in `setup'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:92:in `run'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:78:in `run'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:45:in `invoke'
	from /Users/connor/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.4.4/exe/rspec:4:in `<top (required)>'
	from /Users/connor/.rvm/gems/ruby-2.3.0/bin/rspec:22:in `load'
	from /Users/connor/.rvm/gems/ruby-2.3.0/bin/rspec:22:in `<main>'
	from /Users/connor/.rvm/gems/ruby-2.3.0/bin/ruby_executable_hooks:15:in `eval'
	from /Users/connor/.rvm/gems/ruby-2.3.0/bin/ruby_executable_hooks:15:in `<main>'
```

connorshea commented:
@bsedat fixed it! It fails a bunch of the tests, but it runs. Had to update the TwoFactorBackupable specs as well.

connorshea commented:
Tests pass with attr_encrypted 1.4.0, but not 2.0.0. Based on [the Changelog](https://github.com/attr-encrypted/attr_encrypted/blob/master/CHANGELOG.md#200) and [update section of the README](https://github.com/attr-encrypted/attr_encrypted#deprecations) it looks like devise-two-factor isn't passing the mode to attr_encrypted, and therefore using the default. The default changed in 2.0.0, and encryptor is giving us errors in the test suite about the key not being long enough.

I think this is something we should probably take the time to fix. We can merge this PR with just the Devise 4 fixes and then open a separate one for updating attr_encrypted. devise-two-factor 3.0.0 shouldn't be released until attr_encrypted is updated if we choose to do that, as it will cause breaking changes.

@bsedat do you (or anyone else at tinfoil) have the ability to spend some time on this? I'd rather not handle security issues myself, seeing as I'm a designer and not really experienced with this kind of thing.

bsedat commented:
I'm totally fine limiting the scope of this project to updating Devise compatibility and locking down our supported version of attr_encrypted for now. I doubt anyone at Tinfoil has bandwidth this week though but we'll probably get to it soon-ish.

Thank you again for sticking through this!

bsedat commented:
We're just figuring out how we want to merge this in while still preserving Ruby 2.0 compatibility (at least in a branch). Should have it in shortly (and probably some fixes for attr_encrypted compatibility following hopefully shortly as well).

connorshea commented:
Awesome! Thank you for taking the time to look into it :)

bsedat commented:
I've added this commit (minus the attr_encrypted fixes which are in #73) to a new development branch `devise-4`. Once devise 4.0 is fully released we can merge that branch into master and release a new version of this gem.

Thanks very much again @connorshea for pushing development of this library forward!

connorshea commented:
:+1: happy to be of service!

connorshea commented:
I feel rude for asking, but I would appreciate a mention in the merge commit :sweat_smile:

bsedat commented:
You're totally correct. I've reauthored that commit to you.

connorshea commented:
Thank you :)

Feel free to contact me when Rails 5 and Devise 4 come out, and I'll see if I can test v3.0 with GitLab as well. There's currently [an issue open for upgrading to Rails 5](https://gitlab.com/gitlab-org/gitlab-ce/issues/14286) and I've been collecting all the gems we'll need to update. Glad there's one less potential blocker to worry about now!
